examples/gno.land/r/gnoland/faucet/admin.gno

```diff
@@ -5,44 +5,83 @@ import (
 	"std"
 )
 
-func AdminSetInPause(inPause bool) error {
+func AdminSetInPause(inPause bool) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
 	gInPause = inPause
-	return nil
+	return ""
 }
 
-func AdminSetMessage(message string) error {
+func AdminSetMessage(message string) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
 	gMessage = message
-	return nil
+	return ""
 }
 
-func AdminSetPerTransferSend(send std.Coins) error {
+func AdminSetTransferLimit(amount int64, denom string) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
-	gPerTransferSend = send
-	return nil
+	gLimit = std.Coins{std.Coin{Denom: denom, Amount: amount}}
+	return ""
 }
 
-func AdminSetAdminAddr(addr std.Address) error {
+func AdminSetAdminAddr(addr std.Address) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
 	gAdminAddr = addr
-	return nil
+	return ""
 }
 
-func AdminSetControllerAddr(addr std.Address) error {
+func AdminAddController(addr std.Address) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
-	gControllerAddr = addr
-	return nil
+
+	k := -1
+	for _, v := range gControllers {
+		if v == addr {
+			return addr.String() + " exists, no need to add."
+		}
+	}
+	for i, v := range gControllers {
+		if v == "" {
+
+			k = i
+			break
+		}
+	}
+	if k < 0 {
+		return "can not add more controllers than allowed"
+	}
+
+	gControllers[k] = addr
+
+	return ""
+}
+
+func AdminRemoveController(addr std.Address) string {
+	if err := assertIsAdmin(); err != nil {
+		return err.Error()
+	}
+
+	removed := false
+	for i, v := range gControllers {
+		if v == addr {
+
+			gControllers[i] = ""
+			removed = true
+		}
+	}
+
+	if removed == false {
+		return addr.String() + " is not on the controller list"
+	}
+	return ""
```

Comment:
```suggestion
func AdminRemoveController(addr std.Address) string {
	if err := assertIsAdmin(); err != nil {
		return err.Error()
	}

	for i, v := range gControllers {
		if v == addr {
			gControllers[i] = ""
			return ""
		}
	}

	return addr.String() + " is not on the controller list"
```

Reply:
We don't want to remove the logic of "remove:= false"  and loop through the array. 

Because we need to check all slots in an array, we can not assume to the Address being removed is the only copy, even though the single entry is guarded at AddController API. 

These contract APIs should be written in a way that does not depend on the assumption of other methods of implementation and check all the states.

Reply:
Ok, please use the `avl` package or create a dedicated package to manage a new datatype.

It's an official package that will be used for educational purposes. It should not contain too many workarounds to make a storage-efficient limited map with a slice.

Reply:
Sure, I will change it.